### PR TITLE
Migrate NVText Stemming APIs to pylibcudf

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Apart of #15162 
## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/nvtext/stemmer.pyx [hunk 1]
```diff
@@ -27,43 +22,34 @@ class LetterType(IntEnum):
...
 def is_letter(Column strings,
               object ltype,
               size_type index):
-    cdef column_view c_strings = strings.view()
-    cdef letter_type c_ltype = <letter_type>(
-        <underlying_type_t_letter_type> ltype
+    return Column.from_pylibcudf(
+        nvtext.stemmer.is_letter(
+            strings.to_pylibcudf(mode="read"),
+            True if ltype==LetterType.VOWEL else False,
```
Comment (bdice) [inline python/cudf/cudf/_lib/nvtext/stemmer.pyx:39]:
```suggestion
            ltype == LetterType.VOWEL,
```

File: python/cudf/cudf/_lib/nvtext/stemmer.pyx [hunk 2]
```diff
@@ -27,43 +22,34 @@ class LetterType(IntEnum):
...
                     object ltype,
                     Column indices):
-    cdef column_view c_strings = strings.view()
-    cdef column_view c_indices = indices.view()
-    cdef letter_type c_ltype = <letter_type>(
-        <underlying_type_t_letter_type> ltype
+    return Column.from_pylibcudf(
+        nvtext.stemmer.is_letter(
+            strings.to_pylibcudf(mode="read"),
+            True if ltype==LetterType.VOWEL else False,
```
Comment (bdice) [inline python/cudf/cudf/_lib/nvtext/stemmer.pyx:52]:
```suggestion
            ltype == LetterType.VOWEL,
```

File: python/pylibcudf/pylibcudf/nvtext/__init__.pxd [hunk 3]
```diff
@@ -16,4 +17,5 @@ __all__ = [
     "minhash",
     "ngrams_tokenize",
     "normalize",
+    "stemmer"
```
Comment (bdice) [inline python/pylibcudf/pylibcudf/nvtext/__init__.pxd:20]:
```suggestion
    "stemmer",
```

File: python/pylibcudf/pylibcudf/nvtext/stemmer.pyx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        Input strings
+    check_vowels : bool
+        If true, the check is for vowels. Otherwise the check is
+        for consonants.
+    indices : Union[Column, size_type]
+        The character position(s) to check in each string
+    Returns
+    -------
+    Column
+        New strings columns of normalized strings.
```
Comment (bdice) [inline python/pylibcudf/pylibcudf/nvtext/stemmer.pyx:40]: This sounds incorrect.

File: python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    return pa.array(arr)
+
+
+@pytest.mark.parametrize("check_vowels", [True, False])
+@pytest.mark.parametrize("indices", [[3, 1, 4], 1])
+def test_is_letter(input_col, check_vowels, indices):
+    def is_letter(s, i, check):
+        vowels = ["a", "e", "i", "o", "u", "y"]
+        if check:
+            if s[i] in vowels:
```
Comment (bdice) [inline python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py:21]: Remove the "return True" / etc. as we already have the result.
```suggestion
            return s[i] in vowels
```

File: python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                return True
+            return False
+        if s[i] not in vowels:
```
Comment (bdice) [inline python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py:24]: Similar simplification here.

**Review by bdice — CHANGES_REQUESTED** [review]
A few suggestions. Generally looks good.

File: python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd [hunk 7]
```diff
@@ -9,8 +9,8 @@ from pylibcudf.libcudf.types cimport size_type
 
 cdef extern from "nvtext/stemmer.hpp" namespace "nvtext" nogil:
     ctypedef enum letter_type:
-        CONSONANT 'nvtext::letter_type::CONSONANT'
-        VOWEL 'nvtext::letter_type::VOWEL'
+        CONSONANT "nvtext::letter_type::CONSONANT"
```
[the remark this answers is not in the document]
Comment (Matt711) [reply python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd]: Thanks for catching that!

File: python/pylibcudf/pylibcudf/nvtext/stemmer.pyx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    Parameters
+    ----------
+    input : Column
+        Input strings
+    check_vowels : bool
+        If true, the check is for vowels. Otherwise the check is
+        for consonants.
+    indices : Union[Column, size_type]
+        The character position(s) to check in each string
+    Returns
```
Comment (Matt711) [inline python/pylibcudf/pylibcudf/nvtext/stemmer.pyx:37]:
```suggestion
        The character position(s) to check in each string

    Returns
```

Comment (bdice) [reply python/pylibcudf/pylibcudf/nvtext/stemmer.pyx:37]: Yes, please insert a line here.

File: python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+import pyarrow as pa
+import pylibcudf as plc
+import pytest
+from utils import assert_column_eq
+
+
+@pytest.fixture(scope="module")
+def input_col():
+    arr = ["trouble", "toy", "sygyzy"]
```
Comment (bdice) [inline python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py:11]: Not a bug, really, but the word is syzygy. Maybe we can fix this in C++ tests too.

09716570b4661b35a8e2a8be336d21ee

```suggestion
    arr = ["trouble", "toy", "syzygy"]
```

File: python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd [hunk 10]
```diff
@@ -1,16 +1,17 @@
...
 from pylibcudf.libcudf.column.column cimport column
 from pylibcudf.libcudf.column.column_view cimport column_view
 from pylibcudf.libcudf.types cimport size_type
 
 
 cdef extern from "nvtext/stemmer.hpp" namespace "nvtext" nogil:
-    ctypedef enum letter_type:
-        CONSONANT 'nvtext::letter_type::CONSONANT'
-        VOWEL 'nvtext::letter_type::VOWEL'
+    cpdef enum class letter_type(bool):
```
Comment (bdice) [inline python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd:12]: There isn't a storage class (`bool`) defined for this enum class in the C++ code... we should probably define that at the C++ level, so it looks like this:

https://github.com/rapidsai/cudf/blob/02ee8199a7703eafe8aa061213423eecf27757ec/cpp/include/cudf/types.hpp#L118-L121

**Review by bdice — COMMENTED** [review]
Requesting a couple of minor C++ improvements to pair with this PR.

File: python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py (hunk 9 above)
Comment (Matt711) [reply python/pylibcudf/pylibcudf/tests/test_nvtext_stemmer.py:11]: Ha! TIL

File: python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd (hunk 10 above)
Comment (Matt711) [reply python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd:12]: I wonder if I can just get rid of the `bool` instead

Comment (Matt711) [reply python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd]: Yes, I think that shouldn't have been there, given the enum class is defined without it

Comment (bdice) [reply python/pylibcudf/pylibcudf/libcudf/nvtext/stemmer.pxd]: Seems fine to me.

**Review by bdice — APPROVED** [review]

**Pull request merged**